src/test/java/picard/sam/markduplicates/MarkDuplicatesTagRepresentativeReadIndexTester.java

```diff
@@ -0,0 +1,126 @@
+ * @author hogstrom@broadinstitute.org
+ */
+public class MarkDuplicatesTagRepresentativeReadIndexTester extends AbstractMarkDuplicatesCommandLineProgramTester {
+
+    //final public Map<String, String> readToRepReadMap = new HashMap<>(); // map to contain representative read names for each record
+    final public Map<Integer, Integer> expectedRepresentativeIndexMap = new HashMap<>();
+    final public Map<String, Integer> expectedSetSizeMap = new HashMap<>();
+    public boolean testRepresentativeReads = false;
+
+    public MarkDuplicatesTagRepresentativeReadIndexTester() {
+
+        addArg("TAGGING_POLICY=All");
+        addArg("TAG_DUPLICATE_SET_MEMBERS=true");
+    }
+
+    @Override
+    protected CommandLineProgram getProgram() { return new MarkDuplicates(); }
+
+    @Override
+    public void test() {
+        try {
+            updateExpectedDuplicationMetrics();
+            // Read the output and check the duplicate flag
+            int outputRecords = 0;
+            int indexInFile = 0;
+            final SamReader reader = SamReaderFactory.makeDefault().open(getOutput());
+            for (final SAMRecord record : reader) {
+                outputRecords++;
+
+                final String key = samRecordToDuplicatesFlagsKey(record);
+                if (!this.duplicateFlags.containsKey(key)) {
+                    System.err.println("DOES NOT CONTAIN KEY: " + key);
+                }
+                Assert.assertTrue(this.duplicateFlags.containsKey(key));
+                final boolean value = this.duplicateFlags.get(key);
+                this.duplicateFlags.remove(key);
+                if (value != record.getDuplicateReadFlag()) {
+                    System.err.println("Mismatching read:");
+                    System.err.print(record.getSAMString());
+                }
+                Assert.assertEquals(record.getDuplicateReadFlag(), value);
+                if (testRepresentativeReads) {
+                    if (expectedRepresentativeIndexMap.containsKey(indexInFile) && expectedSetSizeMap.containsKey(record.getReadName())){
+                        Assert.assertEquals(record.getAttribute("DI"), expectedRepresentativeIndexMap.get(indexInFile));
+                        Assert.assertEquals(record.getAttribute("DS"), expectedSetSizeMap.get(record.getReadName()));
+                    }
+                }
+                indexInFile+=1;
+            }
+            CloserUtil.close(reader);
+
+            // Ensure the program output the same number of records as were read in
+            Assert.assertEquals(outputRecords, this.getNumberOfRecords(), ("saw " + outputRecords + " output records, vs. " + this.getNumberOfRecords() + " input records"));
+
+            // Check the values written to metrics.txt against our input expectations
+            final MetricsFile<DuplicationMetrics, Comparable<?>> metricsOutput = new MetricsFile<DuplicationMetrics, Comparable<?>>();
```

Comment:
I think you don't need to explicitly state your template classes, and can just do,
new MetricsFile<>();